src/sync.h

```diff
@@ -103,6 +103,9 @@ class LOCKABLE AnnotatedMixin : public PARENT
     }
 
     using UniqueLock = std::unique_lock<PARENT>;
+#ifdef __clang__
+    const AnnotatedMixin& operator!() const { return *this; }
+#endif // __clang__
```

Comment:
Here and in `StdMutex` - I find this very confusing because it would end up with `m` being the same as `!m`. It is a huge hack imposed by clang. I would suggest to add some comment to explain what is that. Something like `Required by the clang thread safety instrumentation so that we can declare that a mutex should not be held when a function is called - EXCLUSIVE_LOCKS_REQUIRED(!m) (aka "negative capabilities")`.

The example in https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#mutex-h also has some comment.

PS I wonder what would `EXCLUSIVE_LOCKS_REQUIRED(!!m)` do. :eyes:

Reply:
Done in #19249.